mypy/test/testutil.py

```diff
@@ -0,0 +1,83 @@
+from typing import Iterator, List, Tuple
+import time
+import os
+import sys
+import tempfile
+from contextlib import contextmanager
+from threading import Thread
+from unittest import TestCase, main, skipUnless
+from mypy import util
+from mypy.build import random_string
+
+
+WIN32 = sys.platform.startswith("win")
+
+
+def lock_file(filename: str, duration: float) -> Thread:
+    """Open a file and keep it open in a background thread for duration sec."""
+    def _lock_file() -> None:
+        with open(filename):
+            time.sleep(duration)
+    t = Thread(target=_lock_file, daemon=True)
+    t.start()
+    return t
+
+
+@skipUnless(WIN32, "only relevant for Windows")
+class WindowsReplace(TestCase):
+    tmpdir = tempfile.TemporaryDirectory(prefix='mypy-test-',
+                                         dir=os.path.abspath('tmp-test-dirs'))
+    # Choose timeout value that would ensure actual wait inside util.replace is close to timeout
+    timeout = 0.0009 * 2 ** 10
+    short_lock = timeout / 4
+    long_lock = timeout * 2
+
+    threads = []  # type: List[Thread]
```

Comment:
My idea was that the entire file locking arrangement is shared across all the `test*` methods of `WIndowsReplace` test case. Specifically, I want to wait for all the file locks to expire at the end, rather than wait for the file lock in each test to expire. The former is much more efficient: while the other tests are running, old file locks will naturally use up part (or even all) of their duration, so the wait at the end will be much shorter. If we adopt this approach, I need `threads` to be class-level since `WindowsReplace` is instantiated as many times as there are `test*` methods in it. It also means I must not delete the `tmpdir` until the end of the entire test case, i.e., until `tearDownClass` rather than instance-level `tearDown`. Is there a disadvantage to this approach?

Note: I temporarily reduced the number of `test*` metods to 1 due to issues with `xdist` sending different tests to different workers (and thus defeating the entire mechanism I described in the previous paragraph). But this is a horrible arrangement; once I figure out how to resolve the `xdist` issue, I definitely want to follow the standard practice of keeping each test in its own method.